### PR TITLE
Manually merge 3.x into master

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

File: src/Admin/BreadcrumbsBuilder.php [hunk 1]
```diff
@@ -122,7 +122,7 @@ public function buildBreadcrumbs(
         }
 
         if ('list' === $action) {
-            $menu->setUri(null);
+            $menu->setUri('');
```
Comment (VincentLanglet) [inline src/Admin/BreadcrumbsBuilder.php:125]: null is better, the change was made on master

File: tests/App/Datagrid/ProxyQuery.php [hunk 2]
```diff
@@ -76,8 +79,8 @@ public function getUniqueParameterId(): int
         return 1;
     }
 
-    public function entityJoin(array $associationMappings): string
+    public function entityJoin(array $associationMappings): array
```
Comment (VincentLanglet) [inline tests/App/Datagrid/ProxyQuery.php:82]: It's returning a string I think.

Comment (greg0ire) [reply tests/App/Datagrid/ProxyQuery.php:82]: Ah I hesitated here too

File: src/Admin/AdminInterface.php [hunk 3]
```diff
@@ -72,6 +72,9 @@ public function attachAdminClass(FieldDescriptionInterface $fieldDescription): v
 
     public function getDatagrid(): DatagridInterface;
 
+    // NEXT_MAJOR: uncomment this method in 4.0
+    //public function getPagerType(): string;
```
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:76]: You lost the @method and didn't uncomment this.

File: src/Admin/BaseFieldDescription.php [hunk 4]
```diff
@@ -432,7 +432,7 @@ private function callCachedGetter(object $object, string $fieldName, array $para
         }
 
         if ('call' === self::$fieldGetters[$getterKey]['method']) {
-            return $object->__call($fieldName, $parameters);
+            return $object->{$fieldName}(...$parameters);
```
Comment (VincentLanglet) [inline src/Admin/BaseFieldDescription.php:435]: I don't know if we should promote __call or `{}`.
The `__call` change was made on master I think.

**Comment by greg0ire** [comment]
> I did #6302 in the same time joy

What a coincidence

> Let's see if one of us can have a green build.

With your help I think I just got one :)

File: src/Datagrid/PagerInterface.php [hunk 5]
```diff
@@ -15,6 +15,12 @@
 
 /**
  * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * NEXT_MAJOR: Remove these comments and uncomment corresponding methods.
+ *
+ * @method int  getPage()
+ * @method bool isLastPage()
+ * @method int  getNbResults()
```
Comment (VincentLanglet) [inline src/Datagrid/PagerInterface.php:23]: They now exists

File: tests/App/Datagrid/ProxyQuery.php [hunk 6]
```diff
@@ -78,6 +81,6 @@ public function getUniqueParameterId(): int
 
     public function entityJoin(array $associationMappings): string
     {
-        return 'alias';
+        throw new \BadMethodCallException('Not implemented.');
```
Comment (VincentLanglet) [inline tests/App/Datagrid/ProxyQuery.php:84]: We could have kept the `return 'alias'` then.

**Review by VincentLanglet — COMMENTED** [review]
> > I did #6302 in the same time joy
>
> What a coincidence

Sure !

I needed the `getPagerType` method in the PagerInterface of master, that's why I made the merge.

I have few feedback.

**Comment by greg0ire** [comment]
@VincentLanglet looks like we did not obtain the same result:

```diff
diff --git a/phpstan.neon.dist b/phpstan.neon.dist
index 4e75f11f..0cce313b 100644
--- a/phpstan.neon.dist
+++ b/phpstan.neon.dist
@@ -2,7 +2,7 @@ includes:
     - phpstan-baseline.neon
 
 parameters:
-    level: 5
+    level: 4
 
     paths:
         - src
diff --git a/src/Admin/AdminInterface.php b/src/Admin/AdminInterface.php
index 44ddd1a9..95051094 100644
--- a/src/Admin/AdminInterface.php
+++ b/src/Admin/AdminInterface.php
@@ -72,7 +72,8 @@ interface AdminInterface extends AccessRegistryInterface, FieldDescriptionRegist
 
     public function getDatagrid(): DatagridInterface;
 
-    public function getPagerType(): string;
+    // NEXT_MAJOR: uncomment this method in 4.0
+    //public function getPagerType(): string;
 
     /**
      * Set base controller name.
diff --git a/src/Admin/BaseFieldDescription.php b/src/Admin/BaseFieldDescription.php
index 82e1453e..1fc7dc57 100644
--- a/src/Admin/BaseFieldDescription.php
+++ b/src/Admin/BaseFieldDescription.php
@@ -432,7 +432,7 @@ abstract class BaseFieldDescription implements FieldDescriptionInterface
         }
 
         if ('call' === self::$fieldGetters[$getterKey]['method']) {
-            return $object->__call($fieldName, $parameters);
+            return $object->{$fieldName}(...$parameters);
         }
 
         return $object->{$fieldName};
diff --git a/src/Datagrid/PagerInterface.php b/src/Datagrid/PagerInterface.php
index 15dacf62..44de1c8c 100644
--- a/src/Datagrid/PagerInterface.php
+++ b/src/Datagrid/PagerInterface.php
@@ -15,6 +15,12 @@ namespace Sonata\AdminBundle\Datagrid;
 
 /**
  * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * NEXT_MAJOR: Remove these comments and uncomment corresponding methods.
+ *
+ * @method int  getPage()
+ * @method bool isLastPage()
+ * @method int  getNbResults()
  */
 interface PagerInterface
 {
@@ -77,4 +83,8 @@ interface PagerInterface
     public function isLastPage(): bool;
 
     public function getNbResults(): int;
+
+//
+//    NEXT_MAJOR: uncomment this method in 4.0
+//    public function getPage(): int;
 }
diff --git a/src/Datagrid/ProxyQueryInterface.php b/src/Datagrid/ProxyQueryInterface.php
index 27855d7a..d901e52f 100644
--- a/src/Datagrid/ProxyQueryInterface.php
+++ b/src/Datagrid/ProxyQueryInterface.php
@@ -50,10 +50,8 @@ interface ProxyQueryInterface
      * NEXT_MAJOR: Remove this method.
      *
      * @deprecated since sonata-project/admin-bundle 3.x, to be removed in 4.0.
-     *
-     * @return mixed
      */
-    public function getSingleScalarResult();
+    public function getSingleScalarResult(): ?int;
 
     public function setFirstResult(?int $firstResult): self;
 
diff --git a/src/Security/Handler/AclSecurityHandler.php b/src/Security/Handler/AclSecurityHandler.php
index 13b5a970..851659bc 100644
--- a/src/Security/Handler/AclSecurityHandler.php
+++ b/src/Security/Handler/AclSecurityHandler.php
@@ -19,6 +19,7 @@ use Symfony\Component\Security\Acl\Domain\RoleSecurityIdentity;
 use Symfony\Component\Security\Acl\Domain\UserSecurityIdentity;
 use Symfony\Component\Security\Acl\Exception\AclNotFoundException;
 use Symfony\Component\Security\Acl\Exception\NotAllAclsFoundException;
+use Symfony\Component\Security\Acl\Model\AclInterface;
 use Symfony\Component\Security\Acl\Model\MutableAclInterface;
 use Symfony\Component\Security\Acl\Model\MutableAclProviderInterface;
 use Symfony\Component\Security\Acl\Model\ObjectIdentityInterface;
@@ -155,7 +156,7 @@ final class AclSecurityHandler implements AclSecurityHandlerInterface
         $this->deleteAcl($objectIdentity);
     }
 
-    public function getObjectAcl(ObjectIdentityInterface $objectIdentity): ?MutableAclInterface
+    public function getObjectAcl(ObjectIdentityInterface $objectIdentity)
     {
         try {
             $acl = $this->aclProvider->findAcl($objectIdentity);
@@ -181,16 +182,38 @@ final class AclSecurityHandler implements AclSecurityHandlerInterface
         return $acls;
     }
 
-    public function addObjectOwner(MutableAclInterface $acl, ?UserSecurityIdentity $securityIdentity = null): void
+    /**
+     * NEXT_MAJOR: change signature to `addObjectOwner(MutableAclInterface $acl, ?UserSecurityIdentity $securityIdentity = null): void`.
+     *
+     * @param MutableAclInterface $acl
+     */
+    public function addObjectOwner(AclInterface $acl, ?UserSecurityIdentity $securityIdentity = null): void
     {
+        // NEXT_MAJOR: remove `if` condition
+        if (!$acl instanceof MutableAclInterface) {
+            throw new \TypeError(sprintf(
+                'Argument 1 passed to "%s()" must implement "%s".',
+                __METHOD__,
+                MutableAclInterface::class
+            ));
+        }
         if (false === $this->findClassAceIndexByUsername($acl, $securityIdentity->getUsername())) {
             // only add if not already exists
             $acl->insertObjectAce($securityIdentity, \constant("$this->maskBuilderClass::MASK_OWNER"));
         }
     }
 
-    public function addObjectClassAces(MutableAclInterface $acl, array $roleInformation = []): void
+    /**
+     * Add the object class ACE's to the object ACL.
+     *
+     * NEXT_MAJOR: change signature to `addObjectClassAces(MutableAclInterface $acl, array $roleInformation = []): void`.
+     *
+     * @param MutableAclInterface $acl
+     */
+    public function addObjectClassAces(AclInterface $acl, array $roleInformation = []): void
     {
+        // NEXT_MAJOR: remove `assert` statement
+        \assert($acl instanceof MutableAclInterface);
         $builder = new $this->maskBuilderClass();
 
         foreach ($roleInformation as $role => $permissions) {
@@ -219,13 +242,25 @@ final class AclSecurityHandler implements AclSecurityHandlerInterface
         }
     }
 
-    public function createAcl(ObjectIdentityInterface $objectIdentity): MutableAclInterface
+    /**
+     * NEXT_MAJOR: change signature to `createAcl(ObjectIdentityInterface $objectIdentity): MutableAclInterface`.
+     *
+     * @return MutableAclInterface
+     */
+    public function createAcl(ObjectIdentityInterface $objectIdentity)
     {
         return $this->aclProvider->createAcl($objectIdentity);
     }
 
-    public function updateAcl(MutableAclInterface $acl): void
+    /**
+     * NEXT_MAJOR: change signature to `updateAcl(MutableAclInterface $acl): void`.
+     *
+     * @param MutableAclInterface $acl
+     */
+    public function updateAcl(AclInterface $acl): void
     {
+        // NEXT_MAJOR: remove `assert` statement
+        \assert($acl instanceof MutableAclInterface);
         $this->aclProvider->updateAcl($acl);
     }
 
@@ -234,7 +269,14 @@ final class AclSecurityHandler implements AclSecurityHandlerInterface
         $this->aclProvider->deleteAcl($objectIdentity);
     }
 
-    public function findClassAceIndexByRole(MutableAclInterface $acl, $role)
+    /**
+     * NEXT_MAJOR: change signature to `findClassAceIndexByRole(MutableAclInterface $acl, string $role): int|string|false`.
+     *
+     * @param MutableAclInterface $acl
+     *
+     * @return array-key|false
+     */
+    public function findClassAceIndexByRole(AclInterface $acl, $role)
     {
         foreach ($acl->getClassAces() as $index => $entry) {
             if ($entry->getSecurityIdentity() instanceof RoleSecurityIdentity && $entry->getSecurityIdentity()->getRole() === $role) {
@@ -245,7 +287,14 @@ final class AclSecurityHandler implements AclSecurityHandlerInterface
         return false;
     }
 
-    public function findClassAceIndexByUsername(MutableAclInterface $acl, $username)
+    /**
+     * NEXT_MAJOR: change signature to `findClassAceIndexByUsername(MutableAclInterface $acl, string $username): int|string|false`.
+     *
+     * @param MutableAclInterface $acl
+     *
+     * @return array-key|false
+     */
+    public function findClassAceIndexByUsername(AclInterface $acl, $username)
     {
         foreach ($acl->getClassAces() as $index => $entry) {
             if ($entry->getSecurityIdentity() instanceof UserSecurityIdentity && $entry->getSecurityIdentity()->getUsername() === $username) {
diff --git a/src/Security/Handler/AclSecurityHandlerInterface.php b/src/Security/Handler/AclSecurityHandlerInterface.php
index ef6df9f1..8bb2fa1a 100644
--- a/src/Security/Handler/AclSecurityHandlerInterface.php
+++ b/src/Security/Handler/AclSecurityHandlerInterface.php
@@ -14,6 +14,7 @@ declare(strict_types=1);
 namespace Sonata\AdminBundle\Security\Handler;
 
 use Symfony\Component\Security\Acl\Domain\UserSecurityIdentity;
+use Symfony\Component\Security\Acl\Model\AclInterface;
 use Symfony\Component\Security\Acl\Model\MutableAclInterface;
 use Symfony\Component\Security\Acl\Model\ObjectIdentityInterface;
 
@@ -48,8 +49,10 @@ interface AclSecurityHandlerInterface extends SecurityHandlerInterface
 
     /**
      * Get the ACL for the passed object identity.
+     *
+     * @return MutableAclInterface|null
      */
-    public function getObjectAcl(ObjectIdentityInterface $objectIdentity): ?MutableAclInterface;
+    public function getObjectAcl(ObjectIdentityInterface $objectIdentity);
 
     /**
      * Find the ACLs for the passed object identities.
@@ -66,28 +69,44 @@ interface AclSecurityHandlerInterface extends SecurityHandlerInterface
     /**
      * Add an object owner ACE to the object ACL.
      *
+     * NEXT_MAJOR: change signature to `addObjectOwner(MutableAclInterface $acl, ?UserSecurityIdentity $securityIdentity = null): void`.
+     *
+     * @param MutableAclInterface $acl
+     *
      * @return void
      */
-    public function addObjectOwner(MutableAclInterface $acl, ?UserSecurityIdentity $securityIdentity = null);
+    public function addObjectOwner(AclInterface $acl, ?UserSecurityIdentity $securityIdentity = null);
 
     /**
      * Add the object class ACE's to the object ACL.
      *
+     * NEXT_MAJOR: change signature to `addObjectClassAces(MutableAclInterface $acl, array $roleInformation = []): void`.
+     *
+     * @param MutableAclInterface $acl
+     *
      * @return void
      */
-    public function addObjectClassAces(MutableAclInterface $acl, array $roleInformation = []);
+    public function addObjectClassAces(AclInterface $acl, array $roleInformation = []);
 
     /**
      * Create an object ACL.
+     *
+     * NEXT_MAJOR: change signature to `createAcl(ObjectIdentityInterface $objectIdentity): MutableAclInterface`
+     *
+     * @return MutableAclInterface
      */
-    public function createAcl(ObjectIdentityInterface $objectIdentity): MutableAclInterface;
+    public function createAcl(ObjectIdentityInterface $objectIdentity);
 
     /**
      * Update the ACL.
      *
+     * NEXT_MAJOR: change signature to `updateAcl(MutableAclInterface $acl): void`
+     *
+     * @param MutableAclInterface $acl
+     *
      * @return void
      */
-    public function updateAcl(MutableAclInterface $acl);
+    public function updateAcl(AclInterface $acl);
 
     /**
      * Delete the ACL.
@@ -97,18 +116,22 @@ interface AclSecurityHandlerInterface extends SecurityHandlerInterface
     /**
      * Helper method to find the index of a class ACE for a role.
      *
+     * NEXT_MAJOR: change signature to `findClassAceIndexByRole(MutableAclInterface $acl, string $role): array-key|false`
+     *
      * @param string $role
      *
-     * @return string|int|false index if found, FALSE if not found
+     * @return array-key|false index if found, FALSE if not found
      */
-    public function findClassAceIndexByRole(MutableAclInterface $acl, $role);
+    public function findClassAceIndexByRole(AclInterface $acl, $role);
 
     /**
      * Helper method to find the index of a class ACE for a username.
      *
+     * NEXT_MAJOR: change signature to `findClassAceIndexByUsername(MutableAclInterface $acl, string $username): array-key|false`
+     *
      * @param string $username
      *
-     * @return string|int|false index if found, FALSE if not found
+     * @return array-key|false index if found, FALSE if not found
      */
-    public function findClassAceIndexByUsername(MutableAclInterface $acl, $username);
+    public function findClassAceIndexByUsername(AclInterface $acl, $username);
 }
diff --git a/tests/App/Datagrid/ProxyQuery.php b/tests/App/Datagrid/ProxyQuery.php
index 76d2296e..7aef3086 100644
--- a/tests/App/Datagrid/ProxyQuery.php
+++ b/tests/App/Datagrid/ProxyQuery.php
@@ -49,7 +49,7 @@ final class ProxyQuery implements ProxyQueryInterface
     /**
      * NEXT_MAJOR: Remove this method.
      */
-    public function getSingleScalarResult()
+    public function getSingleScalarResult(): ?int
     {
         return 0;
     }
@@ -81,6 +81,6 @@ final class ProxyQuery implements ProxyQueryInterface
 
     public function entityJoin(array $associationMappings): string
     {
-        return 'alias';
+        throw new \BadMethodCallException('Not implemented.');
     }
 }

```

**Comment by VincentLanglet** [comment]
You can ignore the `AclSecurityHandler`, I made extra work that now I think about should be done in another PR.

But you should update the level of phpstan to 5 indeed !

File: src/Admin/AdminInterface.php (hunk 3 above)
Comment (greg0ire) [reply src/Admin/AdminInterface.php:76]: I never apply the `NEXT_MAJOR` on merge, and thought the person who would would fix the `@method` issue, so to me it wasn't required.

Comment (VincentLanglet) [reply src/Admin/AdminInterface.php:76]: > I never apply the `NEXT_MAJOR` on merge, and thought the person who would would fix the `@method` issue, so to me it wasn't required.

The person who introduced this was me. That's why I made the change on my PR 😅
I'll made this after your PR then.

File: src/Admin/BaseFieldDescription.php (hunk 4 above)
Comment (greg0ire) [reply src/Admin/BaseFieldDescription.php:435]: Change on 3.x: 6fa2e473c185f1a8d80758dbe0a17299867a05a9

**Comment by VincentLanglet** [comment]
Test are fine on my side if you want @greg0ire https://github.com/sonata-project/SonataAdminBundle/pull/6302

File: src/Admin/BaseFieldDescription.php (hunk 4 above)
Comment (greg0ire) [reply src/Admin/BaseFieldDescription.php:435]: Change on master: 54e9f496440005b14171ea695b9f3998d279025e

**Comment by greg0ire** [comment]
Ok, go ahead and push, you know the latest changes better than I do. Closing in favor of #6302

File: src/Admin/BaseFieldDescription.php (hunk 4 above)
Comment (VincentLanglet) [reply src/Admin/BaseFieldDescription.php:435]: Oh indeed ! weird.

**Pull request closed without merging**